cidc_api/utils/migrations.py

```diff
@@ -0,0 +1,159 @@
+import os
+from typing import Callable, List, NamedTuple
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.orm.session import Session
+from google.cloud import storage
+
+from cidc_api.models import (
+    TrialMetadata,
+    DownloadableFiles,
+    AssayUploads,
+    AssayUploadStatus,
+    ManifestUploads,
+)
+from cidc_api.config.settings import GOOGLE_DATA_BUCKET, GOOGLE_UPLOAD_BUCKET
+from cidc_schemas.migrations import MigrationResult
+from cidc_schemas.prism import _get_uuid_info
+
+
+class PieceOfWork(NamedTuple):
+    do: Callable[[], None]
+    undo: Callable[[], None]
+
+
+class RollbackableQueue:
+    """A collection of reversible pieces-of-work."""
+
+    def __init__(self):
+        self.tasks = []
+
+    def schedule(self, task: PieceOfWork):
+        """Add a task to the task queue."""
+        self.tasks.append(task)
+
+    def run_all(self):
+        """
+        Attempt to run all tasks in the queue, rolling back
+        successfully completed tasks if a subsequent task fails.
+        """
+        for i, task in enumerate(self.tasks):
+            try:
+                task.do()
+            except:
+                for done_task in self.tasks[:i]:
+                    done_task.undo()
+                raise
+
+
+def run_metadata_migration(metadata_migration: Callable[[dict], MigrationResult]):
+    """Migrate trial metadata, upload job patches, and downloadable files according to `metadata_migration`"""
+    session = Session(bind=op.get_bind())
+
+    # Initialize an empty list of pieces of work to do on GCS resources
+    gcs_tasks = RollbackableQueue()
+
+    # Migrate all trial records
+    trials: List[TrialMetadata] = session.query(TrialMetadata).with_for_update().all()
+    for trial in trials:
+        migration = metadata_migration(trial.metadata_json)
+
+        # Update the trial metadata object
+        trial.metadata_json = migration.result
+
+        # Update the relevant downloadable files and GCS objects
+        for old_gcs_uri, artifact in migration.file_updates.items():
+            # Update the downloadable file associated with this blob
+            df: DownloadableFiles = (
+                session.query(DownloadableFiles)
```

Comment:
Isn't there a way to do update right away, without `with_for_update` locking here?
Seems that unlike as with `TrialMetadata` here we don't rely on current prop values from DF record, to set new ones.